Modules/gcmodule.c

```diff
@@ -1506,6 +1506,97 @@ static PyMethodDef GcMethods[] = {
     {NULL,      NULL}           /* Sentinel */
 };
 
+typedef struct {
+    PyObject_HEAD
+    int previous_gc_state;
+    PyGILState_STATE gstate;
+} disabled_object;
+
+
+static void
+disabled_object_dealloc(disabled_object *m_obj)
+{
+    Py_TYPE(m_obj)->tp_free((PyObject*)m_obj);
+}
+
+
+
+static PyObject *
+disabled__enter__method(disabled_object *self, PyObject *args)
+{
+    self->gstate = PyGILState_Ensure();
+    self->previous_gc_state = _PyRuntime.gc.enabled;
+    _PyRuntime.gc.enabled = 0;
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+disabled__exit__method(disabled_object *self, PyObject *args)
+{
+    _PyRuntime.gc.enabled = self->previous_gc_state;
+    PyGILState_Release(self->gstate);
+    Py_RETURN_NONE;
+}
+
+
+
+static struct PyMethodDef disabled_object_methods[] = {
+    {"__enter__",       (PyCFunction) disabled__enter__method,      METH_NOARGS},
+    {"__exit__",        (PyCFunction) disabled__exit__method,       METH_VARARGS},
+    {NULL,         NULL}       /* sentinel */
+};
+
+static PyObject *
+new_thing(PyTypeObject *type, PyObject *args, PyObject *kwdict){
+    disabled_object *self;
+    self = (disabled_object *)type->tp_alloc(type, 0);
+    return (PyObject *) self;
```

Comment:
Out of curiosity, is this any different from a single statement:

    return (PyObject *) type->tp_alloc(type, 0);

Reply:
Not in this case (unless I am missing something) but is the convention when calling `tp_alloc`. Check for example:

https://github.com/python/cpython/blob/master/Modules/itertoolsmodule.c#L37

Reply:
Cool, thanks for the quick answer :)